redis/credentials.py

```diff
@@ -0,0 +1,77 @@
+from typing import Callable, Optional, Union
+
+
+class CredentialProvider:
+    def __init__(
+        self,
+        username: Union[str, None] = "",
+        password: Union[str, None] = "",
+        supplier: Optional[Callable] = None,
+        *supplier_args,
+        **supplier_kwargs,
+    ):
+        """
+        Initialize a new Credentials Provider.
+        :param supplier: a supplier function that returns the username and password.
+                         def supplier(self, arg1, arg2, ...) -> (username, password)
+                         See examples/connection_examples.ipynb
+        :param supplier_args: arguments to pass to the supplier function
+        :param supplier_kwargs: keyword arguments to pass to the supplier function
+        """
+        self._username_ = "" if username is None else username
+        self._password_ = "" if password is None else password
```

Comment:
Shouldn't these be `self._username` and `self._password` if you want to signal these to be "private"? I haven't ever seen `_foo_` as a pattern for that in all of the Python code I've read.

Reply:
I’ll let @chayim answer, he’s the python expert

Reply:
TBC these aren't the changes that concern me. But, yes - I suggested wrapped underscores (a mistake!. Based on the standard, they should be single sided as @akx suggested.

Reply:
changed